src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,166 @@
+  switch (scan_type) {
+    case ScanType::OpEquals:
+      // Note: The unary "+" forces conversion to a function pointer - without it, the return type would be deduced to
+      //       be the type of the first lambda.
+      return +[](const T& lhs, const U& rhs) { return lhs == rhs; };
+    case ScanType::OpNotEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs != rhs; };
+    case ScanType::OpLessThan:
+      return +[](const T& lhs, const U& rhs) { return lhs < rhs; };
+    case ScanType::OpLessThanEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs <= rhs; };
+    case ScanType::OpGreaterThan:
+      return +[](const T& lhs, const U& rhs) { return lhs > rhs; };
+    case ScanType::OpGreaterThanEquals:
+      return +[](const T& lhs, const U& rhs) { return lhs >= rhs; };
+    default:
+      Fail("Invalid Scan type.");
+  }
+}
+}  // namespace
+
+template <typename T>
+void TableScan::_scan_value_segment(const ChunkID chunk_id, const ValueSegment<T>& segment, PosList& pos_list) {
+  const auto& values = segment.values();
+  const auto segment_size = segment.size();
+  const auto predicate = predicate_for_scantype<T>(_scan_type);
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < segment_size; ++chunk_offset) {
+    if (!segment.is_null(chunk_offset) && !variant_is_null(_search_value) &&
+        predicate(values[chunk_offset], get<T>(_search_value))) {
+      pos_list.push_back({chunk_id, chunk_offset});
+    }
+  }
+}
+
+// Represents a ValueID that can be just between two integer numbers. This is used for comparing range scans of
+// ValueIDs in DictionarySegments where the cutoff for the scan does not have a ValueID assigned as it is not contained
+// in the dictionary.
+struct InBetweenValueID {
+  // clang-format off
+
+  friend bool operator==(const ValueID& other, const InBetweenValueID& self) {
+      return other == self.lower && self.is_contained();
+  }
+  friend bool operator!=(const ValueID& other, const InBetweenValueID& self) { return !(other == self); }
+  friend bool operator<(const ValueID& other, const InBetweenValueID& self) { return other < self.lower; }
+  friend bool operator<=(const ValueID& other, const InBetweenValueID& self) { return other < self.upper; }
+  friend bool operator>(const ValueID& other, const InBetweenValueID& self) { return other >= self.upper; }
+  friend bool operator>=(const ValueID& other, const InBetweenValueID& self) { return other >= self.lower; }
+
+  // clang-format on
+
+  // Whether the lower and upper bounds imply that the search value is contained in the segment.
+  bool is_contained() const {
+    return lower != upper;
+  }
+
+  ValueID lower, upper;
+};
+
+template <typename T>
+void TableScan::_scan_dictionary_segment(const ChunkID chunk_id, const DictionarySegment<T>& segment,
+                                         PosList& pos_list) {
+  const auto attribute_vector = segment.attribute_vector();
```

Comment:
Because one returns a `const vector&` and the other a `shared_ptr<vector>` - both functions choose the type that takes them exactly one physical pointer deref away from the vector